packages/common/src/search/searchCatalogs.ts

```diff
@@ -0,0 +1,93 @@
+import { IArcGISContext } from "../ArcGISContext";
+import { Catalog } from "./Catalog";
+import {
+  ICatalogSearchResponse,
+  IHubCatalog,
+  IPagingOptions,
+  IQuery,
+  ISearchResponseHash,
+  ISortOptions,
+} from "./types";
+import { setProp } from "../objects/set-prop";
+
+/**
+ * Given an array of catalog json objects, execute a search on all the collections
+ * in all the catalogs. If a scope exists without any associated collections in the catalog,
+ * the scope will be searched as well.
+ * If passed an IQuery, only collections using the same targetEntity will be searched
+ * If passed a string, a query will be executed on all collections in all catalogs
+ * @param catalogs
+ * @param query
+ * @param options
+ * @param context
+ * @returns
+ */
+
+export async function searchCatalogs(
+  catalogs: IHubCatalog[],
+  query: string | IQuery,
+  options: IPagingOptions & ISortOptions,
+  context: IArcGISContext
+): Promise<ICatalogSearchResponse[]> {
+  // instantiate Catalog instances
+  const instances = catalogs.map((catalog) =>
+    Catalog.fromJson(catalog, context)
+  );
+
+  // Behavior:
+  // Query all the collections in all the catalogs
+  // if there is a scope which does not have a collection, query the scope
+  const promises = instances.map(async (cat) => {
+    // search the collections
+    const collectionPromise = cat
+      .searchCollections(query, options)
+      .then((response) => {
+        return {
+          catalogTitle: cat.title,
+          collectionResults: response,
+        } as ICatalogSearchResponse;
+      });
+    const promiseArray = [collectionPromise];
+    // for this catalog, find any scopes that do not have a collection
+    const collectionTargets = cat.collections
+      .map((collection) => collection.targetEntity)
+      .filter((v, i, self) => self.indexOf(v));
```

Comment:
So this gives us an array of targetEntities with duplicates removed? It took me a while to understand what was going on with the filter part. Maybe a comment would help our future selves?

Reply:
I'll re-work so it's more obvious - this was some generated code, which while correct, is not as clear as...
```
const collectionTargets = mapBy("targetEntity", cat.collections)
     .filter(unique);
```